site/content/blog/identify-ghost-diff-during-kc-reconcile.md

```diff
@@ -0,0 +1,276 @@
+
+```bash
+$ kubectl get pods
+NAME                          READY   STATUS    RESTARTS   AGE
+simple-app-8648457765-8jtpq   1/1     Running   0          56s
+simple-app-8648457765-p5lzp   1/1     Running   0          56s
+```
+
+## Identify ghost diffs exist or not
+
+In our package, we have set `syncPeriod` to 10 min. This means after every 10 min, kapp controller will try to reconcile the package. kapp creates a configmap every time it sees that there are some resources that needs to be redeployed. Thus, if we see new configmaps appearing, it means ghost diffs are being generated. To identify the configmaps related to an installed package, we have to look at the configmap with Installed Package Prefix. Let's check the configmaps.
+
+```bash 
+$ kubectl get configmaps | grep pkg-demo
+pkg-demo-ctrl                     1      1m40s
+pkg-demo-ctrl-change-ndr7w        1      1m40s
+```
+
+Lets do the same after 10 min of package installation, so that reconcilliation cycle would have run once. 
+
+```bash
+$ kubectl get configmaps | grep pkg-demo
+pkg-demo-ctrl                     1      12m
+pkg-demo-ctrl-change-ndr7w        1      12m
+pkg-demo-ctrl-change-t7zgc        1      56s
+```
+We can see that one more configmap is generated. If I check the configmap content, I can see that there has been one update, but I don't know which resource got updated and what is causing this update. 
+
+```bash
+$ kubectl get configmap pkg-demo-ctrl-change-t7zgc -oyaml
+apiVersion: v1
+data:
+  spec: '{"startedAt":"2022-04-14T08:21:28.264641525Z","finishedAt":"2022-04-14T08:21:32.42972746Z","successful":true,"description":"update:
+    Op: 0 create, 0 delete, 1 update, 0 noop, 0 exists / Wait to: 1 reconcile, 0 delete,
+    0 noop","namespaces":["default"]}'
+kind: ConfigMap
+...
+```
+
+As a Package consumer, I can see that there are ghost diff's appearing. 
+
+## Identify actual configuration causing the diffs
+
+To identify what is causing them, we will make a copy of the package. We will modify the deploy section of the Package. It will help us to get the configuration applied by `kapp`. Let's start:
+
+```bash
+$ kubectl get pkg simple-app.corp.com.1.0.0 -oyaml > copy-simple-app-package.yaml
+```
+
+Open copy-simple-app-package.yaml. Remove labels starting with `kapp`. Add the below snippet to the kapp section. Setting the `diff-changes` to true will enable the `kapp` to show changes.
+```
+...
+- kapp:
+    rawOptions:
+    - --diff-changes=true
+...
+```
+
+I would recommend not to tinker with the original package. Hence, let's change the Package version. Lets change it from 1.0.0 to 2.0.0. Specifically, change the `spec.Version` and `metadata.labels.name`. Now, apply this Package in the cluster so that it will be available for install.
```

Comment:
spec.version (lowercase)
name is not under metadata.labels.name

Reply:
Adopted